docs/rfcs/029-getpage-rate-limiting.md

```diff
@@ -0,0 +1,138 @@
+
+### GetPage@LSN Access Pattern
+
+The Compute has its own hierarchy of caches, specifically `shared_buffers` and the `local file cache` (LFC).
+Compute only issues GetPage requests to Pageserver if it encounters a miss in these caches.
+
+If the working set stops fitting into Compute's caches, requests to Pageserver increase sharply -- the Compute starts *thrashing*.
+
+## Motivation
+
+In INC-69, a tenant issued 155k GetPage/second for a period of 10 minutes and 60k GetPage/second for a period of 3h,
+then dropping to ca 18k GetPage/second for a period of 9h.
+
+We noticed this because of an internal GetPage latency SLO burn rate alert, i.e.,
+the request latency profile during this period significantly exceeded what was acceptable according to the internal SLO.
+
+Sadly, we do not have the observability data to determine the impact of this tenant on other tenants on the same tenants.
+
+However, here are some illustrative data points for the 155k period:
+The tenant was responsible for >= 99% of the GetPage traffic and, frankly, the overall activity on this Pageserver instance.
+We were serving pages at 10 GiB/s (`155k x 8 kbyte (PAGE_SZ) per second is 1.12GiB/s = 9.4Gb/s.`)
+The CPU utilization of the instance was 75% user+system.
+Pageserver page cache served 1.75M accesses/second at a hit rate of ca 90%.
+The hit rate for materialized pages was ca. 40%.
+Curiously, IOPS to the Instance Store NVMe were very low, rarely exceeding 100.
+
+The fact that the IOPS were so low / the materialized page cache hit rate was so high suggests that **this tenant's compute's caches were thrashing**.
+The compute was of type `k8s-pod`; hence, auto-scaling could/would not have helped remediate the thrashing by provisioning more RAM.
+The consequence was that the **thrashing translated into excessive GetPage requests against Pageserver**.
+
+My claim is that it was **unhealthy to serve this workload at the pace we did**:
+* it is likely that other tenants were/would have experienced high latencies (again, we sadly don't have per-tenant latency data to confirm this)
+* more importantly, it was **unsustainable** to serve traffic at this pace for multiple reasons:
+    * **predictability of performance**: when the working set grows, the pageserver materialized page cache hit rate drops.
+      At some point, we're bound by the EC2 Instance Store NVMe drive's IOPS limit.
+      The result is an **uneven** performance profile from the Compute perspective.
+
+    * **economics**: Neon currently does not charge for IOPS, only capacity.
+      **We cannot afford to undercut the market in IOPS/$ this drastically; it leads to adverse selection and perverse incentives.**
+      For example, the 155k IOPS, which we served for 10min, would cost ca. 6.5k$/month when provisioned as an io2 EBS volume.
+      Even the 18k IOPS, which we served for 9h, would cost ca. 1.1k$/month when provisioned as an io2 EBS volume.
```

Comment:
> I think there's a point to be made here that we're not trying to be EBS. We say we autoscale storage, and users could assume this to also include storage IO.

Fair point.

Reply:
> Note: Local attached disks are rated for 16k+ IOPS/disk when you get access to the full disk (assuming is4gen/im4gen is the standard), so this would have been quite dangerous if we were unable to use the OS's page cache.

Where is that 16k number from? And what do you mean by "full disk"?

On the i3en.3xlarge, we can definitely get a lot higher IOPS to the local NVMe.

Reply:
```
admin@ip-172-31-13-23:[/mnt/fio]: fio --name pageserver-startup --bs=8k --size=20G  --rw=randread --direct=1 --runtime=10 --time_based=1 --group_reporting=1 --ioengine=io_uring --iodepth=128
pageserver-startup: (g=0): rw=randread, bs=(R) 8192B-8192B, (W) 8192B-8192B, (T) 8192B-8192B, ioengine=io_uring, iodepth=128
fio-3.25
Starting 1 process
Jobs: 1 (f=1): [r(1)][100.0%][r=1359MiB/s][r=174k IOPS][eta 00m:00s]
pageserver-startup: (groupid=0, jobs=1): err= 0: pid=1731284: Mon Dec 18 19:13:05 2023
  read: IOPS=174k, BW=1359MiB/s (1425MB/s)(13.3GiB/10001msec)
    slat (usec): min=2, max=377, avg= 5.08, stdev= 6.22
    clat (usec): min=44, max=3397, avg=730.03, stdev=219.39
     lat (usec): min=61, max=3415, avg=735.21, stdev=219.42
    clat percentiles (usec):
     |  1.00th=[  449],  5.00th=[  498], 10.00th=[  529], 20.00th=[  570],
     | 30.00th=[  603], 40.00th=[  635], 50.00th=[  668], 60.00th=[  717],
     | 70.00th=[  766], 80.00th=[  857], 90.00th=[ 1012], 95.00th=[ 1172],
     | 99.00th=[ 1516], 99.50th=[ 1680], 99.90th=[ 2040], 99.95th=[ 2180],
     | 99.99th=[ 2573]
   bw (  MiB/s): min= 1352, max= 1364, per=100.00%, avg=1360.11, stdev= 3.45, samples=19
   iops        : min=173108, max=174686, avg=174094.32, stdev=441.07, samples=19
  lat (usec)   : 50=0.01%, 100=0.01%, 250=0.01%, 500=5.07%, 750=61.96%
  lat (usec)   : 1000=22.37%
  lat (msec)   : 2=10.47%, 4=0.12%
  cpu          : usr=20.01%, sys=56.30%, ctx=191575, majf=0, minf=265
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.1%
     issued rwts: total=1739945,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=128

Run status group 0 (all jobs):
   READ: bw=1359MiB/s (1425MB/s), 1359MiB/s-1359MiB/s (1425MB/s-1425MB/s), io=13.3GiB (14.3GB), run=10001-10001msec

Disk stats (read/write):
  nvme1n1: ios=1721876/0, merge=0/0, ticks=1229031/0, in_queue=1229030, util=99.11%
```